src/category_theory/limits/shapes/images.lean

```diff
@@ -405,4 +417,109 @@ def im : arrow C ⥤ C :=
+class has_strong_epi_mono_factorisations :=
+(has_fact : Π {X Y : C} (f : X ⟶ Y), strong_epi_mono_factorisation.{v} f)
+
+@[priority 100]
+instance has_images_of_has_strong_epi_mono_factorisations
+  [has_strong_epi_mono_factorisations.{v} C] : has_images.{v} C :=
+{ has_image := λ X Y f,
+  let F' := has_strong_epi_mono_factorisations.has_fact f in
+  { F := F'.to_mono_factorisation,
+    is_image := F'.to_mono_is_image } }
+
+end strong_epi_mono_factorisation
+
+section has_strong_epi_images
+variables (C) [has_images.{v} C]
+
+/-- A category has strong epi images if it has all images and `factor_thru_image f` is a strong
+    epimorphism for all `f`. -/
+class has_strong_epi_images :=
+(strong_factor_thru_image : Π {X Y : C} (f : X ⟶ Y), strong_epi.{v} (factor_thru_image f))
+
+attribute [instance] has_strong_epi_images.strong_factor_thru_image
+end has_strong_epi_images
+
+section has_strong_epi_images
+
+/-- If we constructed our images from strong epi-mono factorisations, then these images are
+    strong epi images. -/
+@[priority 100]
+instance has_strong_epi_images_of_has_strong_epi_mono_factorisations
+  [has_strong_epi_mono_factorisations.{v} C] : has_strong_epi_images.{v} C :=
+{ strong_factor_thru_image := λ X Y f,
+    (has_strong_epi_mono_factorisations.has_fact f).e_strong_epi }
+
+end has_strong_epi_images
+
+section has_strong_epi_images
+variables [has_images.{v} C]
+
+/-- A category with strong epi images has image maps. The construction is taken from Borceux,
+    Handbook of Categorical Algebra 1, Proposition 4.4.5. -/
+@[priority 100]
+instance has_image_maps_of_has_strong_epi_images [has_strong_epi_images.{v} C] :
+  has_image_maps.{v} C :=
+{ has_image_map := λ f g st,
+    let I := image (image.ι f.hom ≫ st.right),
+    I' := image (st.left ≫ factor_thru_image g.hom),
+    upper : strong_epi_mono_factorisation (f.hom ≫ st.right) :=
+    { I := I,
+      e := factor_thru_image f.hom ≫ factor_thru_image (image.ι f.hom ≫ st.right),
+      m := image.ι (image.ι f.hom ≫ st.right),
+      fac := by simp,
+      e_strong_epi := strong_epi_comp _ _,
+      m_mono := by apply_instance },
+    lower : strong_epi_mono_factorisation (f.hom ≫ st.right) :=
+    { I := I',
+      e := factor_thru_image (st.left ≫ factor_thru_image g.hom),
+      m := image.ι (st.left ≫ factor_thru_image g.hom) ≫ image.ι g.hom,
+      fac := by simp [arrow.w],
+      e_strong_epi := by apply_instance,
+      m_mono := mono_comp _ _ },
+    s : I ⟶ I' := is_image.lift upper.to_mono_is_image lower.to_mono_factorisation in
+    { map := factor_thru_image (image.ι f.hom ≫ st.right) ≫ s ≫
+        image.ι (st.left ≫ factor_thru_image g.hom),
+      factor_map' := by erw [←category.assoc, ←category.assoc,
+        is_image.fac_lift upper.to_mono_is_image lower.to_mono_factorisation, image.fac],
+      map_ι' := by erw [category.assoc, category.assoc,
+        is_image.lift_fac upper.to_mono_is_image lower.to_mono_factorisation, image.fac] } }
```

Comment:
Can you put in a separate `let` here for each? It took a long time for me to parse this, before realising it was multiple let statements. (Perhaps I'm completely wrong, but I think this is our convention --- it saves the misapprehension that `I'`, `upper`, `lower` and `s` are fields of some structure.)